// @wiotp/sdk/dist/api/ApiErrors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.handleError = exports.ServiceNotFound = exports.DestinationAlreadyExists = exports.InvalidServiceCredentials = exports.WiotpError = void 0;

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _wrapNativeSuper(Class) { var _cache = typeof Map === "function" ? new Map() : undefined; _wrapNativeSuper = function _wrapNativeSuper(Class) { if (Class === null || !_isNativeFunction(Class)) return Class; if (typeof Class !== "function") { throw new TypeError("Super expression must either be null or a function"); } if (typeof _cache !== "undefined") { if (_cache.has(Class)) return _cache.get(Class); _cache.set(Class, Wrapper); } function Wrapper() { return _construct(Class, arguments, _getPrototypeOf(this).constructor); } Wrapper.prototype = Object.create(Class.prototype, { constructor: { value: Wrapper, enumerable: false, writable: true, configurable: true } }); return _setPrototypeOf(Wrapper, Class); }; return _wrapNativeSuper(Class); }

function isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _construct(Parent, args, Class) { if (isNativeReflectConstruct()) { _construct = Reflect.construct; } else { _construct = function _construct(Parent, args, Class) { var a = [null]; a.push.apply(a, args); var Constructor = Function.bind.apply(Parent, a); var instance = new Constructor(); if (Class) _setPrototypeOf(instance, Class.prototype); return instance; }; } return _construct.apply(null, arguments); }

function _isNativeFunction(fn) { return Function.toString.call(fn).indexOf("[native code]") !== -1; }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var WiotpError =
/*#__PURE__*/
function (_Error) {
  _inherits(WiotpError, _Error);

  function WiotpError(message, cause) {
    var _this;

    _classCallCheck(this, WiotpError);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(WiotpError).call(this, message));
    _this.cause = cause;
    _this.name = _this.constructor.name;
    return _this;
  }

  return WiotpError;
}(_wrapNativeSuper(Error));

exports.WiotpError = WiotpError;

var InvalidServiceCredentials =
/*#__PURE__*/
function (_WiotpError) {
  _inherits(InvalidServiceCredentials, _WiotpError);

  function InvalidServiceCredentials() {
    _classCallCheck(this, InvalidServiceCredentials);

    return _possibleConstructorReturn(this, _getPrototypeOf(InvalidServiceCredentials).apply(this, arguments));
  }

  return InvalidServiceCredentials;
}(WiotpError);

exports.InvalidServiceCredentials = InvalidServiceCredentials;

var DestinationAlreadyExists =
/*#__PURE__*/
function (_WiotpError2) {
  _inherits(DestinationAlreadyExists, _WiotpError2);

  function DestinationAlreadyExists() {
    _classCallCheck(this, DestinationAlreadyExists);

    return _possibleConstructorReturn(this, _getPrototypeOf(DestinationAlreadyExists).apply(this, arguments));
  }

  return DestinationAlreadyExists;
}(WiotpError);

exports.DestinationAlreadyExists = DestinationAlreadyExists;

var ServiceNotFound =
/*#__PURE__*/
function (_WiotpError3) {
  _inherits(ServiceNotFound, _WiotpError3);

  function ServiceNotFound() {
    _classCallCheck(this, ServiceNotFound);

    return _possibleConstructorReturn(this, _getPrototypeOf(ServiceNotFound).apply(this, arguments));
  }

  return ServiceNotFound;
}(WiotpError);

exports.ServiceNotFound = ServiceNotFound;

var handleError = function handleError(err, errorMappings) {
  if (err && err.response && err.response.data && err.response.data.exception && err.response.data.exception.id) {
    if (errorMappings && errorMappings[err.response.data.exception.id]) {
      throw new errorMappings[err.response.data.exception.id](err.response.data.message, err);
    } else {
      throw new WiotpError(err.response.data.message, err);
    }
  } else {
    throw err;
  }
};

exports.handleError = handleError;
var _default = {
  WiotpError: WiotpError,
  InvalidServiceCredentials: InvalidServiceCredentials,
  DestinationAlreadyExists: DestinationAlreadyExists,
  ServiceNotFound: ServiceNotFound
};
exports["default"] = _default;
