// @wiotp/sdk/dist/application/ApplicationConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _BaseConfig2 = _interopRequireDefault(require("../BaseConfig"));

var _loglevel = _interopRequireDefault(require("loglevel"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var YAML = require('yaml');

var fs = require('fs');

var uuidv4 = require('uuid/v4');

var ApplicationConfig =
/*#__PURE__*/
function (_BaseConfig) {
  _inherits(ApplicationConfig, _BaseConfig);

  function ApplicationConfig(identity, auth, options) {
    var _this;

    _classCallCheck(this, ApplicationConfig);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ApplicationConfig).call(this, identity, auth, options)); // Authentication is not supported for quickstart
    // and not required when auth.useLtpa is set

    if (_this.auth != null && !_this.auth.useLtpa) {
      if (!("key" in _this.auth) || _this.auth.key == null) {
        throw new Error("Missing auth.key from configuration");
      }

      if (!("token" in _this.auth) || _this.auth.token == null) {
        throw new Error("Missing auth.token from configuration");
      }
    } // Set defaults for optional configuration


    if (_this.identity == null) {
      _this.identity = {};
    }

    if (!("appId" in _this.identity)) {
      _this.identity.appId = uuidv4();
    }

    if (!("sharedSubscription" in _this.options.mqtt)) {
      _this.options.mqtt.sharedSubscription = false;
    }

    return _this;
  }

  _createClass(ApplicationConfig, [{
    key: "getOrgId",
    value: function getOrgId() {
      if (this.auth == null) {
        return "quickstart";
      } else if (this.auth.key) {
        return this.auth.key.split("-")[1];
      } else {
        return null;
      }
    }
  }, {
    key: "getApiBaseUri",
    value: function getApiBaseUri() {
      return this.auth && this.auth.useLtpa ? "/api/v0002" : "https://".concat(this.getOrgId(), ".").concat(this.options.domain, "/api/v0002");
    }
  }, {
    key: "getClientId",
    value: function getClientId() {
      var clientIdPrefix = "a";

      if (this.sharedSubscription == true) {
        clientIdPrefix = "A";
      }

      return clientIdPrefix + ":" + this.getOrgId() + ":" + this.identity.appId;
    }
  }, {
    key: "getMqttUsername",
    value: function getMqttUsername() {
      return this.auth.key;
    }
  }, {
    key: "getMqttPassword",
    value: function getMqttPassword() {
      return this.auth.token;
    }
  }], [{
    key: "parseEnvVars",
    value: function parseEnvVars() {
      // Auth
      var authKey = process.env.WIOTP_AUTH_KEY || null;
      var authToken = process.env.WIOTP_AUTH_TOKEN || null; // Also support WIOTP_API_KEY / WIOTP_API_TOKEN usage

      if (authKey == null && authToken == null) {
        authKey = process.env.WIOTP_API_KEY || null;
        authToken = process.env.WIOTP_API_TOKEN || null;
      } // Identity


      var appId = process.env.WIOTP_IDENTITY_APPID || uuidv4(); // Options

      var domain = process.env.WIOTP_OPTIONS_DOMAIN || null;
      var logLevel = process.env.WIOTP_OPTIONS_LOGLEVEL || "info";
      var port = process.env.WIOTP_OPTIONS_MQTT_PORT || null;
      var transport = process.env.WIOTP_OPTIONS_MQTT_TRANSPORT || null;
      var caFile = process.env.WIOTP_OPTIONS_MQTT_CAFILE || null;
      var cleanStart = process.env.WIOTP_OPTIONS_MQTT_CLEANSTART || "true";
      var sessionExpiry = process.env.WIOTP_OPTIONS_MQTT_SESSIONEXPIRY || 3600;
      var keepAlive = process.env.WIOTP_OPTIONS_MQTT_KEEPALIVE || 60;
      var sharedSubs = process.env.WIOTP_OPTIONS_MQTT_SHAREDSUBSCRIPTION || "false";
      var verifyCert = process.env.WIOTP_OPTIONS_HTTP_VERIFY || "true"; // String to int conversions

      if (port != null) {
        port = parseInt(port);
      }

      sessionExpiry = parseInt(sessionExpiry);
      keepAlive = parseInt(keepAlive);
      var identity = {
        appId: appId
      };
      var options = {
        domain: domain,
        logLevel: logLevel,
        mqtt: {
          port: port,
          transport: transport,
          cleanStart: ["True", "true", "1"].includes(cleanStart),
          sessionExpiry: sessionExpiry,
          keepAlive: keepAlive,
          sharedSubscription: ["True", "true", "1"].includes(sharedSubs),
          caFile: caFile
        },
        http: {
          verify: ["True", "true", "1"].includes(verifyCert)
        }
      };
      var auth = null; // Quickstart doesn't support auth, so ensure we only add this if it's defined

      if (authToken != null) {
        auth = {
          key: authKey,
          token: authToken
        };
      }

      return new ApplicationConfig(identity, auth, options);
    }
  }, {
    key: "parseConfigFile",
    value: function parseConfigFile(configFilePath) {
      //Example Application Configuration File:

      /*
      identity:
          appId: myApp
      auth:
          key: a-23gh56-sdsdajhjnee
          token: Ab$76s)asj8_s5
      options:
          domain: internetofthings.ibmcloud.com
          logLevel: error|warning|info|debug
          mqtt:
              instanceId: myInstance
              port: 8883
              transport: tcp
              cleanStart: false
              sessionExpiry: 3600
              keepAlive: 60
              caFile: /path/to/certificateAuthorityFile.pem
          http:
              verify: true  
      */
      var configFile = fs.readFileSync(configFilePath, 'utf8');
      var data = YAML.parse(configFile);

      if (!fs.existsSync(configFilePath)) {
        throw new Error("File not found");
      } else {
        try {
          var _configFile = fs.readFileSync(configFilePath, 'utf8');

          var data = YAML.parse(_configFile);
        } catch (err) {
          throw new Error("Error reading device configuration file: " + err.code);
        }
      }

      if ("options" in data & "logLevel" in data["options"]) {
        var validLevels = ["error", "warning", "info", "debug"];

        if (!validLevels.includes(data["options"]["logLevel"])) {
          throw new Error("Optional setting options.logLevel (Currently: " + data["options"]["logLevel"] + ") must be one of error, warning, info, debug");
        }
      } else {
        data["options"]["logLevel"] = _loglevel["default"].GetLogger(data["options"]["logLevel"].toUpperCase());
      }

      return new ApplicationConfig(data['identity'], data['auth'], data['options']);
    }
  }]);

  return ApplicationConfig;
}(_BaseConfig2["default"]);

exports["default"] = ApplicationConfig;
