// @wiotp/sdk/dist/gateway/GatewayClient.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _BaseClient2 = _interopRequireDefault(require("../BaseClient"));

var _GatewayConfig = _interopRequireDefault(require("./GatewayConfig"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var CMD_RE = /^iot-2\/type\/(.+)\/id\/(.+)\/cmd\/(.+)\/fmt\/(.+)$/;

var util = require('util');

var GatewayClient =
/*#__PURE__*/
function (_BaseClient) {
  _inherits(GatewayClient, _BaseClient);

  function GatewayClient(config) {
    var _this;

    _classCallCheck(this, GatewayClient);

    if (!config instanceof _GatewayConfig["default"]) {
      throw new Error("Config must be an instance of GatewayConfig");
    }

    if (config.isQuickstart()) {
      throw new Error('[GatewayClient:constructor] Quickstart not supported in Gateways');
    }

    _this = _possibleConstructorReturn(this, _getPrototypeOf(GatewayClient).call(this, config));

    _this.log.debug("[GatewayClient:constructor] GatewayClient initialized for " + config.getClientId());

    return _this;
  }

  _createClass(GatewayClient, [{
    key: "connect",
    value: function connect() {
      var _this2 = this;

      _get(_getPrototypeOf(GatewayClient.prototype), "connect", this).call(this);

      this.mqtt.on('connect', function () {// This gateway client implemention does not automatically subscribe to any commands!?
      });
      this.mqtt.on('message', function (topic, payload) {
        _this2.log.debug("[GatewayClient:onMessage] Message received on topic : " + topic + " with payload : " + payload);

        var match = CMD_RE.exec(topic);

        if (match) {
          _this2.emit('command', match[1], match[2], match[3], match[4], payload, topic);
        }
      });
    }
  }, {
    key: "publishEvent",
    value: function publishEvent(eventId, format, payload, qos, callback) {
      return this._publishEvent(this.config.identity.typeId, this.config.identity.deviceId, eventId, format, payload, qos, callback);
    }
  }, {
    key: "publishDeviceEvent",
    value: function publishDeviceEvent(typeid, deviceId, eventid, format, payload, qos, callback) {
      return this._publishEvent(typeId, deviceId, eventid, format, payload, qos, callback);
    }
  }, {
    key: "_publishEvent",
    value: function _publishEvent(typeId, deviceId, eventId, format, data, qos, callback) {
      var topic = util.format("iot-2/type/%s/id/%s/evt/%s/fmt/%s", typeId, deviceId, eventId, format);
      qos = qos || 0;

      this._publish(topic, data, qos, callback);

      return this;
    }
  }, {
    key: "subscribeToDeviceCommands",
    value: function subscribeToDeviceCommands(typeId, deviceId, commandId, format, qos, callback) {
      typeId = typeId || '+';
      deviceId = deviceId || '+';
      commandId = commandid || '+';
      format = format || '+';
      qos = qos || 0;
      var topic = "iot-2/type/" + typeId + "/id/" + deviceId + "/cmd/" + commandId + "/fmt/" + format;

      this._subscribe(topic, qos, callback);

      return this;
    }
  }, {
    key: "unsubscribeFromDeviceCommands",
    value: function unsubscribeFromDeviceCommands(typeId, deviceId, commandId, format, callback) {
      typeId = typeId || '+';
      deviceId = deviceId || '+';
      commandId = commandId || '+';
      format = format || '+';
      var topic = "iot-2/type/" + typeId + "/id/" + deviceId + "/cmd/" + commandId + "/fmt/" + format;

      this._unsubscribe(topic, callback);

      return this;
    }
  }, {
    key: "subscribeToCommands",
    value: function subscribeToCommands(commandId, format, qos, callback) {
      commandId = commandId || '+';
      format = format || '+';
      qos = qos || 0;
      var topic = "iot-2/type/" + this.config.identity.typeId + "/id/" + this.config.identity.deviceId + "/cmd/" + commandId + "/fmt/" + format;

      this._subscribe(topic, qos, callback);

      return this;
    }
  }, {
    key: "unsubscribeFromCommands",
    value: function unsubscribeFromCommands(commandId, format, callback) {
      commandId = commandId || '+';
      format = format || '+';
      var topic = "iot-2/type/" + this.config.identity.typeId + "/id/" + this.config.identity.deviceId + "/cmd/" + commandId + "/fmt/" + format;

      this._unsubscribe(topic, callback);

      return this;
    }
  }]);

  return GatewayClient;
}(_BaseClient2["default"]);

exports["default"] = GatewayClient;
